### PR TITLE
Unperformant Stringify Warning - WIP

### LINKED ISSUE #12209: Warn if stringifying an attribute takes too long
With React 16 we don't have an attribute whitelist so both `src={obj}` and `myattribute={obj}` would be valid. The objects get stringified and added as attributes for smooth migration path because a lot of the existing code already depends on this behavior.

There is, however, one pitfall here. Sometimes you made do `<div {...rest}>` and not realize that `rest` includes an object whose stringifying is unusually expensive. For example a deeply nested Immutable Map. Now, this wouldn't produce an error, but it would slow down rendering for no good reason.

We could protect against this by putting `performance.now()` counters around the places where we stringify attributes. If stringification takes more than, say, 2 milliseconds, then something bad is going on, and we should probably warn.

### PR DESCRIPTION
**Pull request by Swieckowski**
Closes #12209. 
`performance.now()` was not supported while I was writing this, so I'm using the less precise `Date.now()`.

I have written a test to see if the warning shows up if something takes too long to stringify, but don't have one for if it does not warn, since I didn't see anyone else test for anything like that.

### CODE REVIEW

**Comment by xilaraux** [comment]
@Swieckowski 
What did you mean by 
> performance.now() was not supported

**Comment by Swieckowski** [comment]
In the simplest terms, if I run yarn test I'll fail a bunch of tests with ` ReferenceError: performance is not defined`.

**Comment by xilaraux** [comment]
I see, thx for explanation.

**Comment by pshrmn** [comment]
Node `8.5` is required for [`performance.now`](https://nodejs.org/api/perf_hooks.html#perf_hooks_performance_now), but `devEngines` specifies that all `8.X` versions are supported. I don't know what the process is for deciding which Node versions to support, so I'm not sure if `8.5` can just be made the minimum version (it looks like `8.9` is LTS).

File: packages/react-dom/src/client/DOMPropertyOperations.js [hunk 1]
```diff
@@ -132,8 +133,13 @@ export function setValueForProperty(
...
       if (value === null) {
         node.removeAttribute(attributeName);
+      } else if (__DEV__) {
+        node.setAttribute(
+          attributeName,
+          stringifyWithPerformanceWarning(value),
+        );
       } else {
-        node.setAttribute(attributeName, '' + (value: any));
+        node.setAttribute(attributeName, (value: any).toString());
```
Comment (viankakrisna) [inline packages/react-dom/src/client/DOMPropertyOperations.js]: isn't this will throw on `undefined` value? we only strictly check for `null` above it.

**Comment by n3tr** [comment]
Can we do check if `performance.now` is available and fallback to `Date.now()` if it's not, like in ReactNativeFrameScheduling.js

https://github.com/facebook/react/blob/2cf9063318014dcc9a4839bb15ebb89a55198fce/packages/react-native-renderer/src/ReactNativeFrameScheduling.js#L12-L17

File: packages/react-dom/src/client/DOMPropertyOperations.js [hunk 2]
```diff
@@ -172,3 +180,17 @@ export function setValueForProperty(
     }
   }
 }
+
+// Only used in DEV, stringifies a value and Warns if it took too long
+const stringifyWithPerformanceWarning = value => {
+  const stringifyStart = Date.now();
+  let attributeValue = (value: any).toString();
```
Comment (raunofreiberg) [inline packages/react-dom/src/client/DOMPropertyOperations.js]: Use `const` over `let`.

File: packages/react-dom/src/client/DOMPropertyOperations.js [hunk 3]
```diff
@@ -172,3 +180,17 @@ export function setValueForProperty(
...
+  const stringifyEnd = Date.now();
+
+  warning(
+    stringifyEnd - stringifyStart <= 2,
+    'Stringifying your attribute is causing perfomance issues',
```
Comment (raunofreiberg) [inline packages/react-dom/src/client/DOMPropertyOperations.js]: Typo: pefomance => performance.

File: packages/react-dom/src/client/DOMPropertyOperations.js [hunk 4]
```diff
@@ -160,10 +166,12 @@ export function setValueForProperty(
     let attributeValue;
     if (type === BOOLEAN || (type === OVERLOADED_BOOLEAN && value === true)) {
       attributeValue = '';
+    } else if (__DEV__) {
+      attributeValue = stringifyWithPerformanceWarning(value);
     } else {
       // `setAttribute` with objects becomes only `[object]` in IE8/9,
       // ('' + value) makes it output the correct toString()-value.
-      attributeValue = '' + (value: any);
+      attributeValue = (value: any).toString();
```
Comment (raunofreiberg) [inline packages/react-dom/src/client/DOMPropertyOperations.js]: Why was this changed? The comment above no longer reflects the actual code. Aren't we concerned about IE8/9 anymore? If not, then the comment should also be updated, I guess.

Comment (probablyup) [reply packages/react-dom/src/client/DOMPropertyOperations.js]: https://reactjs.org/blog/2016/01/12/discontinuing-ie8-support.html

Just IE9+ it seems

File: packages/react-dom/src/__tests__/DOMPropertyOperations-test.js [hunk 5]
```diff
@@ -155,5 +155,43 @@ describe('DOMPropertyOperations', () => {
       expect(container.firstChild.getAttribute('value')).toBe('foo');
       expect(container.firstChild.value).toBe('foo');
     });
+
+    it('should warn if custom attributes take too long to stringify', () => {
+      const container = document.createElement('div');
+      const attributeValue = {foo: 'bar'};
+      attributeValue.toString = function() {
```
[the remark this answers is not in the document]
Comment (aweary) [reply packages/react-dom/src/__tests__/DOMPropertyOperations-test.js]: Yeah, let's go ahead and go with mocking `performance` instead of actually taking too much time.

File: packages/react-dom/src/client/DOMPropertyOperations.js (hunk 1 above)
Comment (aweary) [reply packages/react-dom/src/client/DOMPropertyOperations.js]: Yeah, @Swieckowski please revert this. We were intentionally not using `toString`

File: packages/react-dom/src/client/DOMPropertyOperations.js (hunk 4 above)
Comment (aweary) [reply packages/react-dom/src/client/DOMPropertyOperations.js]: @Swieckowski please revert his back as well.

File: packages/react-dom/src/client/DOMPropertyOperations.js [hunk 6]
```diff
@@ -172,3 +180,17 @@ export function setValueForProperty(
     }
   }
 }
+
+// Only used in DEV, stringifies a value and Warns if it took too long
+const stringifyWithPerformanceWarning = value => {
```
Comment (aweary) [inline packages/react-dom/src/client/DOMPropertyOperations.js]: For consistency let's move this to the top of the file, and then only define it in DEV

```js

let stringifyWithPerformanceWarning;

if (__DEV__) {
  stringifyWithPerformanceWarning = function(value) {
   // ...
  }
}
``

File: packages/react-dom/src/client/DOMPropertyOperations.js (hunk 2 above)
Comment (aweary) [reply packages/react-dom/src/client/DOMPropertyOperations.js]: Please use `'' + value` instead of `toString` here as well.

File: packages/react-dom/src/client/DOMPropertyOperations.js [hunk 7]
```diff
@@ -172,3 +180,17 @@ export function setValueForProperty(
...
+  const stringifyStart = Date.now();
```
Comment (aweary) [inline packages/react-dom/src/client/DOMPropertyOperations.js]: We want to use `performance.now` if it exists, and fallback to `Date.now` if it doesn't. Lets do the same thing we do here: https://github.com/facebook/react/blob/master/packages/react-native-renderer/src/ReactNativeFrameScheduling.js#L12-L17

File: packages/react-dom/src/client/DOMPropertyOperations.js (hunk 3 above)
Comment (aweary) [reply packages/react-dom/src/client/DOMPropertyOperations.js]: The messaging here isn't clear enough. Let's also pass the attribute name to `stringifyWithPerformanceWarning`, improve the wording, and also include a component stack. You can get a component stack by importing and using `getCurrentFiberStackAddendum` from `ReactDebugCurrentFiber`.

Lets make it look like:

```js
warning(
  stringifyEnd - stringifyStart <= 2,
  'The attribute `%s` took more than 2 milliseconds to stringify. This usually means you provided a large object ' +
  'as the value for a DOM attribute, which can lead to performance issues.%s',
  attributeName,
  getCurrentFiberStackAddendum(),
)
```

File: packages/react-dom/src/client/DOMPropertyOperations.js (hunk 7 above)
Comment (Swieckowski) [reply packages/react-dom/src/client/DOMPropertyOperations.js]: Thanks a lot for the review, I'll get on it sometime soon!

**Comment by gaearon** [comment]
One more concern is that `performance.now()` itself might add overhead. We'd want to measure the impact of this on a large tree.

**Comment by hassaans** [comment]
We'll only be showing the warning in development mode. Performance drops in development should be insignificant.
